<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Outputs - pipeline-call-mtSNV</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../css/version-select.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Outputs";
        var mkdocs_page_input_path = "outputs.md";
        var mkdocs_page_url = null;
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href=".." class="icon icon-home"> pipeline-call-mtSNV
        </a>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../overview/">Overview</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../how-to-run/">How To Run</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../flow-diagram/">Flow Diagram</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../pipeline-steps/">Pipeline Steps</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../inputs/">Inputs</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="./">Outputs</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#primary-outputs">Primary outputs</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#intermediate-outputs">Intermediate outputs</a>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../testing-and-validation/">Testing and Validation</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../references/">References</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../discussions/">Discussions</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../contributors/">Contributors</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../please-see-list-of-contributors-at-github/">Please see list of Contributors at GitHub.</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../license/">License</a>
                </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">pipeline-call-mtSNV</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Outputs</li>
    <li class="wy-breadcrumbs-aside">
          <a href="https://github.com/uclahs-cds/pipeline-call-mtSNV/blob/2082080b59f56da2be032b1fa236fa8e1e6f0fd6/README.md" class="icon icon-github"> Edit on GitHub</a>
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h2 id="outputs">Outputs</h2>
<blockquote>
<p><strong>Note:</strong> All primary alignment (BAM) and variant call (VCF) outputs are indexed, with checksums generated for both the outputs and their index files</p>
</blockquote>
<h3 id="primary-outputs">Primary outputs</h3>
<table>
<thead>
<tr>
<th style="text-align: left;">Process</th>
<th style="text-align: left;">Output</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">align_mtDNA_MToolBox</td>
<td style="text-align: left;"><code>*.bam</code></td>
<td style="text-align: left;">Aligned, sorted, mitochondrial reads in BAM format</td>
</tr>
<tr>
<td style="text-align: left;">align_mtDNA_MToolBox</td>
<td style="text-align: left;"><code>*mt-classification-best-results.csv</code></td>
<td style="text-align: left;">Contains the best haplogroup prediction for each sequence</td>
</tr>
<tr>
<td style="text-align: left;">align_mtDNA_MToolBox</td>
<td style="text-align: left;"><code>*prioritized_variants.txt</code></td>
<td style="text-align: left;">Contains annotation only for prioritized variants for each sample analyzed, sorted by increasing nucleotide variability</td>
</tr>
<tr>
<td style="text-align: left;">align_mtDNA_MToolBox</td>
<td style="text-align: left;"><code>*summary.txt</code></td>
<td style="text-align: left;">Summary of selected options. Includes predicted haplogroups, total and prioritized variants, coverage of reconstructed genomes, count of homoplasmic and heteroplasmic variants</td>
</tr>
<tr>
<td style="text-align: left;">convert_mitoCaller2VCF</td>
<td style="text-align: left;"><code>*.vcf.gz</code></td>
<td style="text-align: left;">MitoCaller variant calls in VCF format</td>
</tr>
<tr>
<td style="text-align: left;">call_heteroplasmy</td>
<td style="text-align: left;"><code>*.tsv</code></td>
<td style="text-align: left;"><strong>[Paired mode only]</strong> A TSV table showing differences between normal and tumor genotypes, along with variant allele frequencies (VAF). <strong>Note:</strong> Heteroplasmy sites appear in the file, only if the VAF difference between normal and tumor samples exceeds 20%. if no somatic event is detected in the tumor sample, the TSV output will contain only the header line.</td>
</tr>
<tr>
<td style="text-align: left;">call_heteroplasmy</td>
<td style="text-align: left;"><code>*.tsv.*</code></td>
<td style="text-align: left;"><strong>[Paired mode only]</strong> Checksum for generated TSV file</td>
</tr>
</tbody>
</table>
<h3 id="intermediate-outputs">Intermediate outputs</h3>
<table>
<thead>
<tr>
<th style="text-align: left;">Process</th>
<th style="text-align: left;">Output</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: left;">extract_mtDNA_BAMQL</td>
<td style="text-align: left;"><code>*.bam</code></td>
<td style="text-align: left;"><strong>[BAM input only]</strong> Outputs BAM file with only mitochondrial reads</td>
</tr>
<tr>
<td style="text-align: left;">extract_mtDNA_SAMtools</td>
<td style="text-align: left;"><code>*.bam</code></td>
<td style="text-align: left;"><strong>[CRAM input only]</strong> Outputs BAM file with only mitochondrial reads</td>
</tr>
<tr>
<td style="text-align: left;">align_mtDNA_MToolBox</td>
<td style="text-align: left;"><code>OUT_*/</code></td>
<td style="text-align: left;">This folder contains additional intermediate files. Description of the contents can be found <a href="https://github.com/mitoNGS/MToolBox/wiki/Output-files">here</a></td>
</tr>
<tr>
<td style="text-align: left;">align_mtDNA_MToolBox</td>
<td style="text-align: left;"><code>*logassemble.txt</code></td>
<td style="text-align: left;">The log file of the assembleMTgenome.py script</td>
</tr>
<tr>
<td style="text-align: left;">align_mtDNA_MToolBox</td>
<td style="text-align: left;"><code>*processed-fastq.tar.gz</code></td>
<td style="text-align: left;">Compressed FASTQ files generated from BAM input files</td>
</tr>
<tr>
<td style="text-align: left;">align_mtDNA_MToolBox</td>
<td style="text-align: left;"><code>*.conf</code></td>
<td style="text-align: left;">Configuration file listing parameters used in an MToolBox run</td>
</tr>
<tr>
<td style="text-align: left;">align_mtDNA_MToolBox</td>
<td style="text-align: left;"><code>*sample.vcf</code></td>
<td style="text-align: left;">Contains mitochondrial variant positions against reference genome</td>
</tr>
<tr>
<td style="text-align: left;">call_mtSNV_mitoCaller</td>
<td style="text-align: left;"><code>*.tsv</code></td>
<td style="text-align: left;">Contains mtDNA variants (i.e., homoplasmies and heteroplasmies)</td>
</tr>
<tr>
<td style="text-align: left;">call_heteroplasmy</td>
<td style="text-align: left;"><code>*unfiltered.tsv</code></td>
<td style="text-align: left;"><strong>[Paired mode only]</strong> Unfiltered tsv table showing differences in the normal genotype vs tumor genotype.</td>
</tr>
<tr>
<td style="text-align: left;">call_heteroplasmy</td>
<td style="text-align: left;"><code>*.pl.programinfo</code></td>
<td style="text-align: left;"><strong>[Paired mode only]</strong> The log file generated for the heteroplasmy call process</td>
</tr>
</tbody>
</table>
<hr />
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../inputs/" class="btn btn-neutral float-left" title="Inputs"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../testing-and-validation/" class="btn btn-neutral float-right" title="Testing and Validation">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
        <span>
          <a href="https://github.com/uclahs-cds/pipeline-call-mtSNV" class="fa fa-github" style="color: #fcfcfc"> GitHub</a>
        </span>
    
    
      <span><a href="../inputs/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../testing-and-validation/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../js/version-select.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
